src/core/column/sumprod.h

```diff
@@ -0,0 +1,101 @@
+//------------------------------------------------------------------------------
+// Copyright 2022 H2O.ai
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//------------------------------------------------------------------------------
+#ifndef dt_COLUMN_SUM_h
+#define dt_COLUMN_SUM_h
+#include "column/virtual.h"
+#include "stype.h"
+
+namespace dt {
+
+  template <typename T, bool SUM>
+  class SumProd_ColumnImpl : public Virtual_ColumnImpl {
+  private:
+    Column col_;
+    bool is_grouped_;
+    Groupby gby_;
+
+  public:
+    SumProd_ColumnImpl(Column &&col, bool is_grouped, const Groupby &gby)
+      :Virtual_ColumnImpl(gby.size(), col.stype()),
+       col_(std::move(col)),
+       is_grouped_(is_grouped),
+       gby_(gby)
+    {
+      xassert(col_.can_be_read_as<T>());
+    }
+
+
+    bool get_element(size_t i, T* out) const override {
+      T result = !SUM; 
+      T value;     
+      size_t i0, i1;
+      gby_.get_group(i, &i0, &i1);
+      if (is_grouped_){
+        bool isvalid = col_.get_element(i, &value);
+        if (isvalid){
+          if (SUM) {            
+            result = static_cast<T>(i1-i0) * value;
+          } else {
+            for (size_t i = i0; i < i1; ++i) {
+              result *= value;
+            }
+          }
+        }
+      } else {
+        for (size_t i = i0; i < i1; ++i) {
+          bool isvalid = col_.get_element(i, &value);
+          if (isvalid){
+            if (SUM) {
+              result += value;
+            } else {
+              result *= value;
+            }          
+          }
+        }
```

Comment:
Ok, so you see that this loop is not parallelized at all. So for the case when there is only one group (ungrouped frame), we calculate sum and product in a single thread. While the frame method `.sum()` is fully parallel.

Reply:
would it be ok then to implement the parallelised version for non groups? or it is not worth copying? do we need to point users to the Frame method, if they are looking for fast aggregations on non-grouped operations?

Reply:
Yes, we need to do something about it as a part of #3081. Here I suggest we only implement what is required for this PR and check that performance of `dt.sum()` is not affected (even if it is not good comparing to `Frame.sum()`.

Reply:
first image is the current implementation in `main`:
![Screenshot_7](https://user-images.githubusercontent.com/25709809/205279673-e5f99ec2-d424-4ee1-acfb-f06ae2a1526c.png)

the second image is the `PR`:

![Screenshot_8](https://user-images.githubusercontent.com/25709809/205280681-2fa399a3-3d3a-496e-836a-282db8fd69fa.png)

Reply:
Yeah, so performance of `dt.sum()` for this PR is comparable to what we currently have in `main`. At the same time, for groupby we are slightly better than pandas, but with no groupby we are worse. Am I correct with interpreting your numbers?

Reply:
Yes you are right. Hopefully we can improve the performance for both grouped and ungrouped. Pypolars is much faster on both counts.

Reply:
We should be good in the case when the number of groups is `>=` the number of cores, and those groups are comparable in their sizes. Otherwise, we are very far from being optimal. 

What I wonder though, is why pandas is so good on ungrouped frames... I thought pandas is single-threaded, just like what we do in this case.

Reply:
I think somehow numpy makes the sum computation really fast... How I do not know

Reply:
for prod, datatable groupby is slower than pandas: 
![Screenshot_11](https://user-images.githubusercontent.com/25709809/205435665-f278f797-1d8d-47af-b865-2f90df67ea7e.png)